feat: add correct k-induction algorithm.

bollu opened the pull request:
Insight: recall that correctness proof increased on thenumber of boolean
outputs where it produced true. The new algorithm ensures that this
invariant is maintained. In the kth step, it shows that either
$\texttt{safe}(s) \implies \texttt{safe}(s + 1) \lor (\texttt{safe}(s) \land \texttt{safe}(s+1) \implies \texttt{safe}(s+2)) \lor (safe(s) \land safe(s+1) \land \dots \land safe(s + k) \implies safe(s+k+1)$ ,
as well as $\texttt{safe}(s0) \land \texttt{safe}(s0 + 1) \land \dots  \texttt{safe}(s0 + k)$.

The fact that we keep the previous inductive invariants around
makes the solver only try to proceed in the cases where depth $k$
was insufficient, thereby ensuring that we always incraese the
number of states we are proving to be \texttt{safe}, or finding a cex thereof.
Thus, termination is guaranteed.

Also recall that the space of boolean functions has a trivial upper
bound, which this process will eventually reach (thanks @hargoniX).

bollu committed:
chore: it gets more mysterious

bollu committed:
chore: add more examples to see WTF is going on

bollu committed:
feat: add better circuit printing to understand what's going on.

bollu committed:
feat: add better circuit printing

bollu committed:
feat: add explanation of automata

bollu committed:
chore: write wrong version, will fix now

bollu committed:
feat: add abstractions to cleanup the circuit building for k-induction

bollu committed:
feat: change circuit building in terms of vars

bollu committed:
chore: disable specialization for the full circuit related tactic

alexkeizer commented:
Mysterious indeed, you've made me curious what this PR is actually about :rofl:

bollu committed:
feat: establish inductive invariant for Test2500.lean

Insight: recall that correctness proof increased on thenumber of boolean
outputs where it produced true. The new algorithm ensures that this
invariant is maintained. In the kth step, it shows that either
safe(s) => safe(s + 1) \/ (safe(s)/\ safe(s+1) => safe(s+2)) \/ ,
as well as safe(s0) /\ safe(s0 + 1) /\ ... safe(s0 + k).

The fact that we keep the previous inductive invariants around
makes the solver only try to proceed in the cases where depth $k$
was insufficient, thereby ensuring that we always incraese the
number of states we are proving to be safe, or finding a cex thereof.
Thus, termination is guaranteed.

Also recall that the space of boolean functions has a trivial upper
bound, which this process will eventually reach (thanks @hargonix).

bollu committed:
feat: fixup tests

bollu committed:
chore: remove .DS_Store

bollu committed:
feat: add runner.py

bollu committed:
feat: add runner that also runs bv_mba as well as bv_circuit_automata
